snapshot/pkg/volume/openebs/snapshot.go

```diff
@@ -0,0 +1,209 @@
+	if snapshotData == nil || snapshotData.Spec.OpenEBSSnapshot == nil {
+		return nil, false, fmt.Errorf("failed to retrieve Snapshot spec")
+	}
+
+	snapshotID := snapshotData.Spec.OpenEBSSnapshot.SnapshotID
+	glog.V(1).Infof("received describe request on snapshot:%v", snapshotID)
+
+	// TODO implement snapshot-info based on snapshotID
+	resp, err := h.SnapshotInfo(snapshotData.Spec.PersistentVolumeRef.Name, snapshotID)
+
+	if err != nil {
+		glog.Errorf("failed to describe snapshot:%v", snapshotID)
+	}
+
+	glog.V(1).Infof("snapshot details:%v", string(resp))
+
+	if len(snapshotData.Status.Conditions) == 0 {
+		return nil, false, fmt.Errorf("No status condtions in VoluemSnapshotData for openebs snapshot type")
+	}
+
+	lastCondIdx := len(snapshotData.Status.Conditions) - 1
+	retCondType := crdv1.VolumeSnapshotConditionError
+
+	switch snapshotData.Status.Conditions[lastCondIdx].Type {
+	case crdv1.VolumeSnapshotDataConditionReady:
+		retCondType = crdv1.VolumeSnapshotConditionReady
+	case crdv1.VolumeSnapshotDataConditionPending:
+		retCondType = crdv1.VolumeSnapshotConditionPending
+		// Error out.
+	}
+	retCond := []crdv1.VolumeSnapshotCondition{
+		{
+			Status:             snapshotData.Status.Conditions[lastCondIdx].Status,
+			Message:            snapshotData.Status.Conditions[lastCondIdx].Message,
+			LastTransitionTime: snapshotData.Status.Conditions[lastCondIdx].LastTransitionTime,
+			Type:               retCondType,
+		},
+	}
+	return &retCond, true, nil
+}
+
+// FindSnapshot finds a VolumeSnapshot by matching metadata
+func (h *openEBSPlugin) FindSnapshot(tags *map[string]string) (*crdv1.VolumeSnapshotDataSource, *[]crdv1.VolumeSnapshotCondition, error) {
+	glog.Infof("FindSnapshot by tags: %#v", *tags)
+
+	// TODO: Implement FindSnapshot
+	return &crdv1.VolumeSnapshotDataSource{
+		OpenEBSSnapshot: &crdv1.OpenEBSVolumeSnapshotSource{
+			SnapshotID: "",
+		},
+	}, nil, nil
+}
+
+// SnapshotRestore restore to any created snapshot
+func (h *openEBSPlugin) SnapshotRestore(snapshotData *crdv1.VolumeSnapshotData, pvc *v1.PersistentVolumeClaim, pvName string, _ map[string]string) (*v1.PersistentVolumeSource, map[string]string, error) {
+	if snapshotData == nil || snapshotData.Spec.OpenEBSSnapshot == nil {
+	}
+
+	// restore snapshot to a PV
+	snapshotID := snapshotData.Spec.OpenEBSSnapshot.SnapshotID
+
+	_, err := h.RestoreSnapshot(pvc.Name, snapshotID)
+
+	if err != nil {
+		glog.Errorf("snapshot :%v restore failed, err:%v", snapshotID, err)
+		return nil, nil, fmt.Errorf("failed to restore %s, err: %v", snapshotID, err)
+	}
+
+	glog.V(1).Infof("snapshot restored successfully to: %v", snapshotID)
+
+	/*pv := &v1.PersistentVolumeSource{
+		ISCSI: &v1.ISCSIVolumeSource{
+			Path:          newSnapPV,
+			EndpointsName: Ep,
+		},
+	}*/
+	return nil, nil, nil
+}
+
+func (h *openEBSPlugin) VolumeDelete(pv *v1.PersistentVolume) error {
```

Comment:
should this be there? is it volume delete or snapshot delete?

Reply:
Its `VolumeDelete` which is created from the snapshot.